### PR TITLE
feat(abtest): support rules properties on exp allocation

### PR DESCRIPTION
**Pull request by qoharu**
## Summary
Adding optional properties parameters to both getExperiments method to support filtered abTesting
- getAllUniverseAllocations(String userId, HashMap<String, String> properties)
- getUniverseAllocation(String userId, String universeId, HashMap<String, String> properties)

### CODE REVIEW

File: src/main/java/com/sayurbox/kale/abtest/ABTestClient.java [hunk 1]
```diff
@@ -2,11 +2,14 @@
...
+import java.util.HashMap;
 import java.util.List;
 
 public interface ABTestClient {
 
     List<GetUniverseAllocationResponse> getAllUniverseAllocations(String userId);
     GetUniverseAllocationResponse getUniverseAllocation(String userId, String universeId);
 
+    List<GetUniverseAllocationResponse> getAllUniverseAllocations(String userId, HashMap<String, String> properties);
+    GetUniverseAllocationResponse getUniverseAllocation(String userId, String universeId, HashMap<String, String> properties);
```
Comment (husniadil) [inline src/main/java/com/sayurbox/kale/abtest/ABTestClient.java:14]: Should we define enum for the properties, since it's based on contract?

Comment (qoharu) [reply src/main/java/com/sayurbox/kale/abtest/ABTestClient.java:14]: Too exposed business context in public :p

File: src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java [hunk 2]
```diff
@@ -39,7 +43,10 @@ public List<GetUniverseAllocationResponse> getFallback() {
     @Override
     protected Request createRequest() {
         String url = String.format(ENDPOINT, baseUrl, userId);
-        return new Request.Builder().post(RequestBody.create(null, new byte[]{})).url(url).build();
+        byte[] body = gson.toJson(properties, propertiesType).getBytes();
+        MediaType contentType = MediaType.parse("application/json");
+
+        return new Request.Builder().post(RequestBody.create(contentType, body)).url(url).build();
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java:49]: RequestBody bs dicreate dr string json aja hrsnya. gk perlu jd bytes. biar lebih singkat. hehehh

`RequestBody.create(contentType, gson.toJson(this.properties))`

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java [hunk 3]
```diff
@@ -40,7 +44,10 @@ public GetUniverseAllocationResponse getFallback() {
     @Override
     protected Request createRequest() {
         String url = String.format(ENDPOINT, baseUrl, userId, universeId);
-        return new Request.Builder().post(RequestBody.create(null, new byte[]{})).url(url).build();
+        byte[] body = gson.toJson(properties, propertiesType).getBytes();
+        MediaType contentType = MediaType.parse("application/json");
+
+        return new Request.Builder().post(RequestBody.create(contentType, body)).url(url).build();
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:50]: sama dgn diatas

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java [hunk 4]
```diff
@@ -5,17 +5,19 @@
 import com.sayurbox.kale.common.client.DataResponse;
 import com.sayurbox.kale.common.command.KaleCommand;
 import io.github.resilience4j.circuitbreaker.CircuitBreaker;
-import okhttp3.OkHttpClient;
-import okhttp3.Request;
-import okhttp3.RequestBody;
-import okhttp3.Response;
+import okhttp3.*;
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:8]: kalo bs gk pake bintang *

File: src/test/java/com/sayurbox/kale/abtest/ABTestClientImplTest.java [hunk 5]
```diff
@@ -4,11 +4,9 @@
 import com.github.tomakehurst.wiremock.junit.WireMockRule;
 import com.sayurbox.kale.abtest.client.GetUniverseAllocationResponse;
 import com.sayurbox.kale.config.KaleConfig;
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.Rule;
-import org.junit.Test;
+import org.junit.*;
```
Comment (andrisasuke) [inline src/test/java/com/sayurbox/kale/abtest/ABTestClientImplTest.java:7]: sama dgn diatas

File: src/main/java/com/sayurbox/kale/abtest/ABTestClient.java [hunk 6]
```diff
@@ -2,11 +2,14 @@
...
 
+import java.util.HashMap;
 import java.util.List;
 
 public interface ABTestClient {
 
     List<GetUniverseAllocationResponse> getAllUniverseAllocations(String userId);
     GetUniverseAllocationResponse getUniverseAllocation(String userId, String universeId);
 
+    List<GetUniverseAllocationResponse> getAllUniverseAllocations(String userId, HashMap<String, String> properties);
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/ABTestClient.java:13]: pake Map<string, string> aja klo di param.  nanti bs diisi HashMap<> pas dipanggil

File: src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java [hunk 7]
```diff
@@ -5,18 +5,20 @@
 import com.sayurbox.kale.common.client.DataResponse;
 import com.sayurbox.kale.common.command.KaleCommand;
 import io.github.resilience4j.circuitbreaker.CircuitBreaker;
-import okhttp3.OkHttpClient;
-import okhttp3.Request;
-import okhttp3.RequestBody;
-import okhttp3.Response;
+import okhttp3.*;
```
[the remark this answers is not in the document]
Comment (qoharu) [reply src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java:8]: sek ganti linter dulu

**Comment by qoharu** [comment]
Comments resolved ya @andrisasuke

**Review by andrisasuke — APPROVED** [review]

**Pull request merged**